client/authority-discovery/src/addr_cache.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+#![warn(missing_docs)]
+
+use rand::{rngs::StdRng, seq::SliceRandom, Rng, SeedableRng};
+use std::{
+	clone::Clone,
+	cmp::{Eq, PartialEq},
+	collections::HashMap,
+	convert::AsRef,
+	hash::Hash,
+};
+
+/// The maximum number of authority connections initialized through the authority discovery module.
+///
+/// In other words the maximum size of the `authority` peer set priority group.
+const MAX_NUM_AUTHORITY_CONN: usize = 10;
+
+/// Cache of Multiaddresses of authority nodes or their sentry nodes.
+//
+// The network peerset interface for priority groups lets us only set an entire group, but we
+// retrieve the addresses of other authorities one by one from the network. To use the peerset
+// interface we need to cache the addresses and always overwrite the entire peerset priority
+// group. To ensure this map doesn't grow indefinitely `purge_old_authorities_from_cache`
+// function is called each time we add a new entry.
+pub(super) struct AddrCache<Id, Addr> {
+	cache: HashMap<Id, Vec<Addr>>,
+
+	/// Random number to seed address selection RNG.
+	///
+	/// A node should only try to connect to a subset of all authorities. To choose this subset one
+	/// uses randomness. The choice should differ between nodes to prevent hot spots, but not within
+	/// each node between each update to prevent connection churn. Thus before each selection we
+	/// seed an RNG with the same seed.
+	rand_addr_selection_seed: u64,
+}
+
+impl<Id, Addr> AddrCache<Id, Addr>
+where
+	Id: Eq + Hash,
+	Addr: Clone + PartialEq + AsRef<[u8]>,
+{
+	pub fn new() -> Self {
+		AddrCache {
+			cache: HashMap::new(),
+			rand_addr_selection_seed: rand::thread_rng().gen(),
+		}
+	}
+
+	pub fn insert(&mut self, id: Id, mut addresses: Vec<Addr>) {
+		if addresses.is_empty() {
+			return;
+		}
+
+		addresses.sort_by(|a, b| {
+			a.as_ref()
+				.partial_cmp(b.as_ref())
```

Comment:
Can you explain why `partial_cmp` is used here instead `cmp`?

AFAIK, u8 implements Ordering and it is infallible. `partial_cmp` should be used for comparisons that can fail for example floats.

Reply:
And why not use `sort_unstable`?

Reply:
Right, using `cmp` instead of `unwrap`ing a `parital_cmp` is a lot better :+1:, thanks!

Switched to `sort_unstable_by` as well.